File: backend/src/ai/trust_engine.py
def calculate_trust_score(data):
    """
    Calculates the trust score based on various factors.

    Args:
        data (dict): A dictionary containing the user's upload history.
            - verifiedUploads (int): Number of verified uploads.
            - correctClassifications (int): Number of correct classifications.
            - municipalityApproved (int): Number of municipality-approved uploads.
            - spamUploads (int): Number of spam uploads.
            - fakeUploads (int): Number of fake uploads.
            - duplicateUploads (int): Number of duplicate uploads.
            - consecutiveValidUploads (int): Number of consecutive valid uploads.

    Returns:
        dict: A dictionary containing the trust score and the reasons.
            - trustScore (int): The calculated trust score.
            - reasons (list): A list of strings explaining the score calculation.
    """
    score = 50
    reasons = ["Initial score: 50"]

    verified_uploads_score = data["verifiedUploads"] * 3
    score += verified_uploads_score
    reasons.append(f"Verified uploads: {verified_uploads_score} points")

    correct_classifications_score = data["correctClassifications"] * 2
    score += correct_classifications_score
    reasons.append(f"Correct classifications: {correct_classifications_score} points")

    municipality_approved_score = data["municipalityApproved"] * 5
    score += municipality_approved_score
    reasons.append(f"Municipality-approved uploads: {municipality_approved_score} points")

    spam_uploads_score = data["spamUploads"] * -8
    score += spam_uploads_score
    reasons.append(f"Spam uploads: {spam_uploads_score} points")

    fake_uploads_score = data["fakeUploads"] * -10
    score += fake_uploads_score
    reasons.append(f"Fake uploads: {fake_uploads_score} points")

    duplicate_uploads_score = data["duplicateUploads"] * -5
    score += duplicate_uploads_score
    reasons.append(f"Duplicate uploads: {duplicate_uploads_score} points")

    if data["consecutiveValidUploads"] >= 5:
        score += 5
        reasons.append("Consecutive valid uploads bonus: 5 points")

    score = max(0, min(100, score))
    reasons.append(f"Final score clamped between 0 and 100: {score}")

    return {"trustScore": score, "reasons": reasons}


if __name__ == "__main__":
    test_data = {
        "verifiedUploads": 10,
        "correctClassifications": 8,
        "municipalityApproved": 5,
        "spamUploads": 2,
        "fakeUploads": 1,
        "duplicateUploads": 3,
        "consecutiveValidUploads": 6
    }

    result = calculate_trust_score(test_data)
    print("\n========== RESULT ==========\n")

    for key, value in result.items():
        if isinstance(value, list):
            print(f"{key}:")
            for item in value:
                print(f"  • {item}")
        else:
            print(f"{key}: {value}")

    print("\n============================")

File: backend/src/ai/trustEngine.js
/**
 * Calculates the trust score for a user based on their upload history and behavior.
 *
 * @param {Object} data - The input data object containing user upload statistics.
 * @param {number} data.verifiedUploads - The number of verified uploads by the user.
 * @param {number} data.correctClassifications - The number of correct classifications by the user.
 * @param {number} data.municipalityApproved - The number of uploads approved by the municipality.
 * @param {number} data.spamUploads - The number of spam uploads by the user.
 * @param {number} data.fakeUploads - The number of fake uploads by the user.
 * @param {number} data.duplicateUploads - The number of duplicate uploads by the user.
 * @param {number} data.consecutiveValidUploads - The number of consecutive valid uploads by the user.
 *
 * @returns {Object} An object containing the calculated trust score and the reasons for the score.
 * @returns {number} trustScore - The calculated trust score for the user (between 0 and 100).
 * @returns {string[]} reasons - An array of strings describing the reasons for the trust score.
 */
function calculateTrustScore(data) {
    // Extract the input data from the data object
    const {
      verifiedUploads,
      correctClassifications,
      municipalityApproved,
      spamUploads,
      fakeUploads,
      duplicateUploads,
      consecutiveValidUploads
    } = data;
  
    // Initialize the trust score to 50
    let trustScore = 50;
    const reasons = [];
  
    // Add points for verified uploads
    trustScore += verifiedUploads * 3;
    reasons.push(`+${verifiedUploads * 3} points for ${verifiedUploads} verified uploads`);
  
    // Add points for correct classifications
    trustScore += correctClassifications * 2;
    reasons.push(`+${correctClassifications * 2} points for ${correctClassifications} correct classifications`);
  
    // Add points for municipality approved uploads
    trustScore += municipalityApproved * 5;
    reasons.push(`+${municipalityApproved * 5} points for ${municipalityApproved} municipality approved uploads`);
  
    // Deduct points for spam uploads
    trustScore -= spamUploads * 8;
    reasons.push(`-${spamUploads * 8} points for ${spamUploads} spam uploads`);
  
    // Deduct points for fake uploads
    trustScore -= fakeUploads * 10;
    reasons.push(`-${fakeUploads * 10} points for ${fakeUploads} fake uploads`);
  
    // Deduct points for duplicate uploads
    trustScore -= duplicateUploads * 5;
    reasons.push(`-${duplicateUploads * 5} points for ${duplicateUploads} duplicate uploads`);
  
    // Add bonus points for consecutive valid uploads
    if (consecutiveValidUploads >= 5) {
      trustScore += 5;
      reasons.push('+5 bonus points for 5 or more consecutive valid uploads');
    }
  
    // Ensure the trust score stays within the range of 0 to 100
    trustScore = Math.max(0, Math.min(100, trustScore));
  
    // Return the calculated trust score and reasons
    return {
      trustScore,
      reasons
    };
  }
  
  // Export the calculateTrustScore function using CommonJS
  module.exports = calculateTrustScore;